text/0000-default-field-values.md

```diff
@@ -0,0 +1,1685 @@
+            // other fields...
+            ..
+        }
+    }
+}
+```
+
+### Defaults for `enum`s
+
+So far, you've have seen how ergonomics were improved for `struct`s.
+For `enum`s, FRU cannot be used as for structs and neither can
+`#[derive(Default)]` be used for each variant or the entire enum.
+Instead, at best, a constructor function can be offered for each variant.
+For a large enum which for example encodes an expression language such
+this could become quite a lot of variants.
+
+By allowing each individual variant of an `enum` to have default values,
+a more profound can achieve a more profound improvement to ergonomics
+than for `struct`s. For example, you may write:
+
+```rust
+enum Foo<'a> {
+    Bar {
+        alpha: u8 = 42,
+        beta: &'a str = "beta's default value",
+    },
+    Baz {
+        gamma: Vec<u8> = Vec::new(),
+        delta: f32
+    }
+}
+```
+
+and then these defaults may be used like so:
+
+```rust
+let a = Foo::Bar { .. };
+let b = Foo::Bar { alpha: 1, .. };
+let c = Foo::Bar { beta: "another beta", .. };
+
+let d = Foo::Baz { delta: 1.0, .. };
+```
+
+## `#[derive(Default)]` in more cases
+
+The `#[derive(..)]` ("custom derive") mechanism works by defining procedural
+*macros*. Because they are macros, these operate on abstract *syntax* and
+don't have more information available. Therefore, when you `#[derive(Default)]`
+on a data type definition as with:
+
+```rust
+#[derive(Default)]
+struct Foo {
+    bar: u8,
+    baz: String,
+}
+```
+
+it only has the immediate "textual" definition available to it.
+
+Because Rust currently does not have an in-language way to define default values,
+you cannot `#[derive(Default)]` in the cases where you are not happy with the
+natural default values that each field's type provides. By extending the syntax
+of Rust such that default values can be provided, `#[derive(Default)]` can be
+used in many more circumstances and thus boilerplate is further reduced.
+
+Currently, `#[derive(Default)]` is not usable for `enum`s. To rectify this
+situation and to allow enum variants to take advantage of their newfound
+ability to have default field values, a `#[default]` attribute is introduced
+that can be attached to variants. This allows you to use `#[derive(Default)]`
+on enums wherefore you can now write:
+
+```rust
+// From rocket:
+
+#[derive(Default)]
+enum Referrer {
+    #[default]
+    NoReferrer,
+    NoReferrerWhenDowngrade,
+    ...
+}
+
+#[derive(Default)]
+enum ExpectCt {
+    #[default]
+    Enforce {
+        duration: Duration = Duration::days(30)
+    },
+    ...
+}
+
+// From git2:
+
+#[derive(Default)]
+enum FileFavor {
+    #[default]
+    Normal,
+    Ours,
+    Theirs,
+    Union,
+}
+```
+
+Note that while `#[default]` makes field defaults more useful and is useful
+on its own, it is technically orthogonal in that it can be extracted out of
+this RFC if need be. However, since it completes the picture, it is proposed
+together with field defaults.
```

Comment:
`Default` is pretty underused in Rust. I think it is either worth soft-deprecating it by not enhancing it further and not recommending its use, or we should figure out what would make it more used (e.g., putting it in the prelude). I feel a bit uncomfortable proposing enhancements without addressing the deeper concern that it is perhaps not the best construct.

Reply:
IMO many of the reasons it is underused are addressed by this PR, such as the ability to easily have defaults that aren't the same as the default values of the field types. I think enhancing it in this way provides a clear benefit, and doesn't contribute to any problems with `Default` today.